core/src/prediction/LocalLinearPredictionStrategy.cpp

```diff
@@ -43,38 +43,41 @@ LocalLinearPredictionStrategy::LocalLinearPredictionStrategy(const Data* origina
         linear_correction_variables(linear_correction_variables){
 };
 
-std::vector<double> LocalLinearPredictionStrategy::predict(size_t sampleID,
-                                                           const std::unordered_map<size_t, double>& weights_by_sampleID,
-                                                           const Observations& observations) {
-  size_t n = observations.get_num_samples();
+std::vector<double> LocalLinearPredictionStrategy::predict(
+    size_t sampleID,
+    const std::unordered_map<size_t, double>& weights_by_sampleID,
+    const Observations& observations) {
   size_t num_variables = linear_correction_variables.size();
-
-  Eigen::MatrixXd weights = Eigen::MatrixXd::Zero(n,n);
-
-  for (auto it = weights_by_sampleID.begin(); it != weights_by_sampleID.end(); ++it){
-    size_t i = it->first;
-    double weight = it->second;
-    weights(i,i) = weight;
+  size_t num_nonzero_weights = weights_by_sampleID.size();
+
+  std::vector<size_t> indices(num_nonzero_weights);
+  Eigen::MatrixXd weights_vec = Eigen::VectorXd::Zero(num_nonzero_weights);
+  {
```

Comment:
braces here are to keep the scope of the loop variable encapsulated

Reply:
Oops, didn't see that this would conflict with the below -- feel free to ignore my nitpick and keep as before :)